### PR TITLE
ARCMWDT: fix assorted build failures

### PR DESCRIPTION
**Pull request by tagunil**
At the moment we have a lot of failed tests on ARC platforms while using ARC MWDT toolchain. This PR is an attempt to fix at least the most obvious build failures before diving into runtime ones.

### CODE REVIEW

File: include/zephyr/posix/posix_types.h [hunk 1]
```diff
@@ -84,11 +84,6 @@ typedef unsigned long useconds_t;
 #endif
 
 /* time related attributes */
-#if !defined(CONFIG_NEWLIB_LIBC) && !defined(CONFIG_ARCMWDT_LIBC)
```
Comment (keith-packard) [inline include/zephyr/posix/posix_types.h:87]: This change doesn't seem to be relevant to making code work with ARCMWDT? It seems to be explicitly skipping this block when using that toolchain, after all?

Comment (tagunil) [reply include/zephyr/posix/posix_types.h:87]: Yeah, so that's a separate commit. It removes a block of code that became unreachable and redundant (see lines 20 to 24 of the same file). Still, if you believe this change is wrong, I'll drop the commit.

Comment (keith-packard) [reply include/zephyr/posix/posix_types.h:87]: Thanks for the explanation. Looks good to me.

**Review by keith-packard — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]
Looks good to me - I guess it should probably be tagged for a back port once 4.0.0 is cut

**Review by stephanosio — DISMISSED** [review]

**Comment by tagunil** [comment]
I've rebased the changes and added one more commit in the middle to prevent native_sim test failures. Specifically, I've made NATIVE_LIBC select TC_PROVIDES_POSIX_C_LANG_SUPPORT_R, as it looks like a reasonable assumption.

File: lib/libc/arcmwdt/libc-hooks.c [hunk 2]
```diff
@@ -71,6 +71,13 @@ int _write(int fd, const char *buf, unsigned int nbytes)
 }
 #endif
 
+#ifndef CONFIG_POSIX_DEVICE_IO
+__weak int fileno(FILE *file)
+{
+	return _fileno(file);
+}
+#endif
```
Comment (cfriedt) [inline lib/libc/arcmwdt/libc-hooks.c:79]: Does something like this work? I guess not many people have access to the arc mwdt toolchain to test 

```suggestion
__weak int _fileno(FILE *file)
{
	return zvfs_fileno(file);
}
```

Comment (tagunil) [reply lib/libc/arcmwdt/libc-hooks.c:79]: Well, we have a `_fileno()` function in our ARC toolchain, but not a `fileno()` function (it's a macro). So my idea was:
- Undefine the macro in any case because it prevents doing anything
- If CONFIG_POSIX_DEVICE_IO is defined, just use `fileno()` from the POSIX subsystem
- Otherwise define `fileno()` as a wrapper around our `_fileno()` to avoid user breakage
I can't see how your proposal solves the issue of missing `fileno()`. But if you have any suggestions regarding the situation I've described above, they will be very welcome.

Comment (cfriedt) [reply lib/libc/arcmwdt/libc-hooks.c:79]: `fileno()` shouldn't be missing if `CONFIG_POSIX_DEVICE_IO` is under.

Comment (tagunil) [reply lib/libc/arcmwdt/libc-hooks.c:79]: Okay, so there are two separate questions: how to fix the issue of missing `fileno()` in the MWDT toolchain and whether we want to tie `_fileno()` to `zvfs_fileno()` unconditionally.
As to the first one, I still think what I did makes sense and I see no other obvious way to fix the issue.
But as to the second one, I'd like to hear your suggestions. Should we do this unconditionally or it should depend on something enabled? CONFIG_ZVFS? CONFIG_FDTABLE? Anything else?

Comment (tagunil) [reply lib/libc/arcmwdt/libc-hooks.c:79]: And, to be honest, I don't understand at all how the `fileno()` and `fdopen()` functions provided by ZVFS could ever be plugged into e.g. `fwrite()` from any embedded libc, either MWDT or Picolibc, because of different interpretations of what a `FILE` is. It might require changing libc integration layers completely.

Comment (cfriedt) [reply lib/libc/arcmwdt/libc-hooks.c:79]: > But as to the second one, I'd like to hear your suggestions. Should we do this unconditionally or it should depend on something enabled? CONFIG_ZVFS? CONFIG_FDTABLE? Anything else?


Good catch, an `ifdef CONFIG_FDTABLE` would probably work

Comment (tagunil) [reply lib/libc/arcmwdt/libc-hooks.c:79]: Well, I've checked the source and found that the MWDT library doesn't use `_fileno()` internally, so there is no way to replace it with the ZVFS one. I'd really like to see someday a nice clean stdio sitting on top of the POSIX layer, but this is not the day yet. As for now, I propose to keep the change as is.

**Comment by tagunil** [comment]
Hi @keith-packard @cfriedt @stephanosio, could you please revisit this PR? The changes from the last review were quite small.

@cfriedt, I agree that the current approach to `fileno()` is not perfect, but the proper integration between libc and the POSIX layer is a much more complex topic than this one. For now I'll be happy to have it buildable.

File: lib/libc/arcmwdt/libc-hooks.c (hunk 2 above)
Comment (keith-packard) [reply lib/libc/arcmwdt/libc-hooks.c:79]: > Does something like this work? I guess not many people have access to the arc mwdt toolchain to test

This is definitely wrong -- the FILE here is from libc, not the (invalid) FILE from zvfs.

**Review by keith-packard — APPROVED** [review]
I don't see any major issues with this PR.

File: lib/libc/arcmwdt/libc-hooks.c (hunk 2 above)
Comment (cfriedt) [reply lib/libc/arcmwdt/libc-hooks.c:79]: > I'd really like to see someday a nice clean stdio sitting on top of the POSIX layer, but this is not the day yet. 

That would be possible if Zephyr decided to become a POSIX OS.

Feel free to request that as an enhancement. I would support it.

Technically, POSIX is already mandatory for some types and Zephyr supports much of what is required already.

Comment (cfriedt) [reply lib/libc/arcmwdt/libc-hooks.c:79]: > This is definitely wrong -- the FILE here is from libc, not the (invalid) FILE from zvfs.

If the libc `FILE*` contained a `void*` representing the Zephyr `fd_entry*`, then that would work.

**Review by cfriedt — APPROVED** [review]

File: Kconfig.zephyr [hunk 3]
```diff
@@ -448,6 +448,7 @@ config CODING_GUIDELINE_CHECK
 config NATIVE_LIBC
 	bool
 	select FULL_LIBC_SUPPORTED
+	select TC_PROVIDES_POSIX_C_LANG_SUPPORT_R
```
Comment (evgeniy-paltsev) [inline Kconfig.zephyr:451]: So previously we didn't provide POSIX reentrant functions in case of native build? Or we provided implementation from Zephyr?

Comment (cfriedt) [reply Kconfig.zephyr:451]: Yes ? Hard to say - I don't often test on native_sim

Comment (tagunil) [reply Kconfig.zephyr:451]: > So previously we didn't provide POSIX reentrant functions in case of native build? Or we provided implementation from Zephyr?

It's complicated. The native implementation always provided the functions but our configuration system was not aware of the fact. It kinda worked, because in-tree users don't bother with selecting the right symbol anyway. But as soon as the libc test got its dependencies fixed, the issue was uncovered.

**Review by stephanosio — APPROVED** [review]
